data/dataset.py

```diff
@@ -0,0 +1,126 @@
+    ):
+        super().__init__()
+
+        self.dataset = dataset
+        self.quantizer = MidiQuantizer(7, 7, 7)
+        self.tokenizer = tokenizer
+        self.masks = AwesomeMasks()
+
+        self.augmentation_probability = augmentation_probability
+        self.masking_probability = masking_probability
+
+    def __len__(self):
+        return len(self.dataset)
+
+    def apply_augmentation(self, record: dict):
+        # shift pitch augmentation
+        if random.random() < self.augmentation_probability:
+            # max shift is octave down or up
+            shift = random.randint(1, 12)
+            record["pitch"] = pitch_shift(record["pitch"], shift)
+
+        # change tempo augmentation
+        if random.random() < self.augmentation_probability:
+            record["dstart"], record["duration"] = change_speed(record["dstart"], record["duration"])
+            # change bins for new dstart and duration values
+            record["dstart_bin"] = np.digitize(record["dstart"], self.quantizer.dstart_bin_edges) - 1
+            record["duration_bin"] = np.digitize(record["duration"], self.quantizer.duration_bin_edges) - 1
+
+        return record
+
+    def apply_masking(self, token_ids: np.ndarray, record: dict):
+        input_token_ids = token_ids.copy()
+        tgt_token_ids = token_ids.copy()
+
+        # masking, adds new key to record dict called masked
+        masked, mask_type = self.masks.apply(record, p=self.masking_probability)
+
+        # source token ids
+        mask_idx = self.tokenizer.token_to_id["<mask>"]
+        input_token_ids[masked] = mask_idx
+
+        # tgt token ids, -100 means loss is not calculated on this token
+        tgt_token_ids[~masked] = -100
+
+        # add mask type token
+        mask_type_idx = self.tokenizer.token_to_id[mask_type]
+        input_token_ids = np.insert(input_token_ids, obj=0, values=mask_type_idx)
+        tgt_token_ids = np.insert(tgt_token_ids, obj=0, values=-100)
+
+        return input_token_ids, tgt_token_ids
+
+    def add_cls_token(self, input_token_ids: np.ndarray, tgt_token_ids: np.ndarray):
+        cls_token = self.tokenizer.token_to_id["<cls>"]
+        input_token_ids = np.insert(input_token_ids, obj=0, values=cls_token)
+        tgt_token_ids = np.insert(tgt_token_ids, obj=0, values=-100)
+
+        return input_token_ids, tgt_token_ids
+
+    def __getitem__(self, index: int) -> dict:
+        record = self.dataset[index]
+
+        filename = record["midi_filename"]
+        # cast list to np array
+        record = {k: np.array(v) for k, v in record.items() if k != "midi_filename"}
+
+        # sanity check, replace NaN with 0
+        if np.any(np.isnan(record["dstart"])):
+            record["dstart"] = np.nan_to_num(record["dstart"], copy=False)
+
+        record = self.apply_augmentation(record)
+        token_ids = self.tokenizer.encode(record)
```

Comment:
Does it actually expand the dataset size? For me it seems like the number of records in the dataset remains the same, only some of them are mannipulated by augmentation. Or was it supposed to work like that? :thinking:

Reply:
The augmentation is random, so every time you call `dataset[idx]` a different augmentation could be applied. My intuition is that running augmentation before train-time would be better, but we'll see 🤔